### PR TITLE
create hadoop module

### PR DESCRIPTION
**Pull request by kunkun-tang**
#1499 makes Hadoop dependencies compileOnly, which caused web-server and exec-server can not initialize Guice successfully if users don't specify their own Hadoop jars in the classpath.
```
Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/hadoop/conf/Configuration
	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
	at java.lang.Class.getDeclaredMethods(Class.java:1975)
	at com.google.inject.internal.ProviderMethodsModule.getProviderMethods(ProviderMethodsModule.java:132)
	at com.google.inject.internal.ProviderMethodsModule.configure(ProviderMethodsModule.java:123)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:349)
	at com.google.inject.spi.Elements.getElements(Elements.java:110)
	at com.google.inject.internal.InjectorShell$Builder.build(InjectorShell.java:138)
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:104)
	at com.google.inject.Guice.createInjector(Guice.java:99)
	at com.google.inject.Guice.createInjector(Guice.java:73)
	at com.google.inject.Guice.createInjector(Guice.java:62)
	at azkaban.soloserver.AzkabanSingleServer.main(AzkabanSingleServer.java:75)
Caused by: java.lang.ClassNotFoundException: org.apache.hadoop.conf.Configuration
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 14 more
```

This patch creates a new HadoopModule, which places Hadoop-related dependencies. we only install this module when Hadoop related injection is needed. Today only HDFS_Storage needs it.

### CODE REVIEW

**Comment by HappyRay** [comment]
Please clarify in the change description that this is only a problem if users don't supply their own Hadoop jars in the classpath.

How did you test?
Is this code path covered in an automated test, unit test?

**Comment by HappyRay** [comment]
Thanks for linking the related changes.

**Comment by kunkun-tang** [comment]
Description Updated. Unit Test should not cover HDFS coverage, and I evaluated the generated jar in staging cluster, and it works. @HappyRay

**Comment by HappyRay** [comment]
Any reason this can't be unit tested?

**Comment by kunkun-tang** [comment]
Yes, it should be unit testable. Added another method by providing Hdfs related props to test Hadoop Injections. @HappyRay

File: azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java [hunk 1]
```diff
@@ -78,6 +66,9 @@ protected void configure() {
   public Class<? extends Storage> resolveStorageClassType() {
     final StorageImplementationType type = StorageImplementationType
         .from(this.config.getStorageImplementation());
+    if (type == StorageImplementationType.HDFS) {
+      install(new HadoopModule(this.props));
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java]: why should we install here? this seems irrelevant to this method's intention.

Comment (kunkun-tang) [reply azkaban-common/src/main/java/azkaban/AzkabanCommonModule.java]: I was considering this question where we should put installModule, but can not find a good approach. First off, install module cannot be put under AzkabanServer class when injector is initialized, because Guice.createInjector can not optionally install a module. The second alternative location is inside `configure` method in AzkabanCommonModule. It will incur duplicate code to resolve storage type, which is same logic inside the method `resolveStorageClassType`. The final alternative is just here. Only when we provide HDFSStorage class, we install this module. I have some idea how we can refactor HDFSStorage bindings, but it will incur a big code change, and we can do it in future.

File: azkaban-common/src/test/java/azkaban/ServiceProviderTest.java [hunk 2]
```diff
@@ -71,7 +76,34 @@ public void testInjections() throws Exception {
     assertThat(injector.getInstance(StorageManager.class)).isNotNull();
     assertThat(injector.getInstance(DatabaseStorage.class)).isNotNull();
     assertThat(injector.getInstance(LocalStorage.class)).isNotNull();
+    assertThatThrownBy(
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/ServiceProviderTest.java]: Is this assert necessary?

File: azkaban-common/src/test/java/azkaban/ServiceProviderTest.java [hunk 3]
```diff
@@ -71,7 +76,34 @@ public void testInjections() throws Exception {
...
+        () -> injector.getInstance(HdfsStorage.class))
+        .isInstanceOf(ConfigurationException.class)
+        .hasMessageContaining("Guice configuration errors");
+
     assertThat(injector.getInstance(Storage.class)).isNotNull();
     assertThat(injector.getInstance(DatabaseOperator.class)).isNotNull();
   }
+
+  @Test
+  public void testHadoopInjection() throws Exception {
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/ServiceProviderTest.java]: Thanks for adding a test.

File: azkaban-common/src/test/java/azkaban/ServiceProviderTest.java (hunk 2 above)
Comment (kunkun-tang) [reply azkaban-common/src/test/java/azkaban/ServiceProviderTest.java]: The ideal test should test out if runtime needs to resolve hadoop jars. However, it cannot be straightforwardly tested. This test can guarantee that Hadoop module is not installed, so I think it has values.
